src/jdk.jlink/share/classes/jdk/tools/jlink/internal/JRTArchive.java

```diff
@@ -0,0 +1,536 @@
+    }
+
+    @Override
+    public Stream<Entry> entries() {
+        try {
+            collectFiles();
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
+        } catch (RuntimeImageLinkException e) {
+            // populate modified files exception
+            throw e.getReason();
+        }
+        return files.stream().map(JRTFile::toEntry);
+    }
+
+    @Override
+    public void open() throws IOException {
+        if (files.isEmpty()) {
+            collectFiles();
+        }
+    }
+
+    @Override
+    public void close() throws IOException {
+        if (!files.isEmpty()) {
+            files.clear();
+        }
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(module, path);
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        return (obj instanceof JRTArchive other &&
+                   Objects.equals(module, other.module) &&
+                   Objects.equals(path, other.path));
+    }
+
+    private void collectFiles() throws IOException {
+        if (files.isEmpty()) {
+            addNonClassResources();
+            // Add classes/resources from image module,
+            // patched with the runtime image link diff
+            files.addAll(ref.open().list()
+                                   .filter(i -> {
+                                       String lookupKey = String.format("/%s/%s", module, i);
+                                       ResourceDiff rd = resDiff.get(lookupKey);
+                                       // Filter all resources with a resource diff
+                                       // that are of kind MODIFIED.
+                                       // Note that REMOVED won't happen since in
+                                       // that case the module listing won't have
+                                       // the resource anyway.
+                                       // Note as well that filter removes files
+                                       // of kind ADDED. Those files are not in
+                                       // the packaged modules, so ought not to
+                                       // get returned from the pipeline.
+                                       return (rd == null ||
+                                               rd.getKind() == ResourceDiff.Kind.MODIFIED);
+                                   })
+                                   .map(s -> {
+                                               String lookupKey = String.format("/%s/%s", module, s);
+                                               return new JRTArchiveFile(JRTArchive.this, s,
+                                                       EntryType.CLASS_OR_RESOURCE,
+                                                       null /* hashOrTarget */,
+                                                       false /* symlink */,
+                                                       resDiff.get(lookupKey));
```

Comment:
Nit: identation within `{...}` better to use consistent spacing (4-space or 8-space is fine).